### ISSUE TITLE
[explicit-function-return-type] allow mixins

### ISSUE THREAD

**Issue by bennypowers** [typescript-eslint/typescript-eslint#2035]
👋 Hello, thank you for publishing this useful package.

This issue is to request a new option on `explicit-function-return-type` which would exempt class mixin functions from the requirement for explicit return types.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/explicit-function-return-type": ["error", { "allowMixins": true }]
  }
}
```

```TS
type Constructor<T ={}> = new (...args: any[]) => T;

export function MyMixin<T extends Constructor>(superclass: T) {
  return class extends superclass {
    mixinProperty = true;
  }
}
```

**Expected Result**
Code passes linting
**Actual Result**
Code errors at linting
**Additional Info**
In most cases, authors would prefer that the return type of a mixin function be inferred, rather than explicitly typing them, which can be fidgity, delicate work with diminishing returns.

**Comment by timvahlbrock** [comment]
I want to propose to keep such an option a little more general, for example like `allowFactories`. This will make the option a little less precise on mixins, but would also support factory functions in general. Implementation could be done somewhat like `allowHigherOrderFunctions` and `allowDirectConstAssertionInArrowFunctions`, so checking if the function immediately returns a class.

**Comment by JoshuaKGoldberg** [comment]
Following up, it's been about 4 years with vvery little activity on this issue. The last attempt to resolve it was ~3 years ago and ended up getting mired in nuances of the scope manager.

We discussed internally and although the `@typescript-eslint/explicit-function-return-type` rule isn't frozen (some of us like it, others prefer [`@typescript-eslint/explicit-module-boundary-types`](https://typescript-eslint.io/rules/explicit-module-boundary-types), it's not something we think is actively in use by enough users to justify fleshing out this style of rare edge case anymore. Mixins as a pattern have become less popular over the years (ACK on the note of expanding this to functions in general). So, I'm going to go ahead and close this out as wontfix. Our recommendation is to use an ESLint disable comment for mixin cases if you want this rule to be enabled.

Cheers, and thanks for the discussion all!
